🐛 Add ability to cache comments count endpoint

lenabaidakova opened the pull request:
closes TryGhost/Team#2094

PR to comments repo https://github.com/TryGhost/comments-ui/pull/12

naz commented on ghost/core/core/shared/config/defaults.json:
```diff
@@ -140,6 +140,9 @@
         "contentAPI": {
             "maxAge": 0
+        },
+        "comments": {
```
I do like the `contentAPI` convention used above. This is an API we are caching, so maybe `commentsAPI` would be a better fit. Also, with the discussion on the issue itself _if_ we end up only caching the comments endpoint long-term, than `commentsCountAPI` would be even better name.

lenabaidakova replied:
Renamed it to `commentsCountAPI`

naz commented on ghost/core/core/server/web/comments/routes.js:
```diff
@@ -1,19 +1,22 @@
 module.exports = function apiRoutes() {
     const router = express.Router('comment api');
 
+    router.use(shared.middleware.cacheControl('public', {maxAge: config.get('caching:comments:maxAge')}));
```
The issue we are solving here is for one specific endpoint and not the whole API. I think we would be most safe to apply this caching rule to the counts endpoint first to unblock the issue and do more research around correct approach for the whole Comments API

lenabaidakova replied:
Added cache just of counts API

naz reviewed (changes requested):
These changes look good. The controller/api level changes are good to go into main I think. The part around setting a custom cache header should to be endpoint-specific.

lenabaidakova commented:
@naz , thank you for explanation 🙌 

Fixed next issues:
- Renamed setting in config to `commentsCountAPI`
- Applied just for one count API
- Added `stale-while-revalidate` option

naz commented on ghost/mw-cache-control/lib/mw-cache-control.js:
```diff
@@ -12,10 +12,18 @@ const isString = require('lodash/isString');
  * @param {'public'|'private'} profile Use "private" if you do not want caching
  * @param {object} [options]
  * @param {number} [options.maxAge] The max-age in seconds to use when profile is "public"
+ * @param {number} [options.staleWhileRevalidate] The stale-while-revalidate in seconds to use when profile is "public"
```
Oh exciting! Nice 😊

naz reviewed (approved):
Yup! All of this looks about right 👌  If I had to be extra nit-picky about it, would separate it into two commits and `staleWhileRevalidate` functionality there with specific explanation for whys and hows 😊

lenabaidakova committed:
🐛 Add ability to cache comments count endpoint

closes TryGhost/Team#2094
- Comment counts request was changed from `post` to `get` to enable request caching.

lenabaidakova committed:
Add ability to pass stale-while-revalidate option to cache control

closes TryGhost/Team#2094
This value can be used for non-crucial data with a `public` option. For example: `public, max-age=1, stale-while-revalidate=9`.
The idea behind this option is that the browser would cache the value for the number of seconds in `max-age` and would use it for the number of seconds in `stale-while-revalidate` until it gets a "validated response" from the server. The behaviour should be almost unnoticeable for the end user but would make a big difference in the amount of requests to server.

lenabaidakova commented:
> Yup! All of this looks about right 👌 If I had to be extra nit-picky about it, would separate it into two commits and `staleWhileRevalidate` functionality there with specific explanation for whys and hows 😊

Fixed :)
